python/heterocl/tvm/build_module.py

```diff
@@ -1,845 +1,856 @@
-"""The build utils in python.
-
-This module provides the functions to transform schedule to
-LoweredFunc and compiled Module.
-"""
-from __future__ import absolute_import as _abs
-import os, subprocess, time, re
-import warnings
-import types
-
-from ._ffi.node import NodeBase, register_node
-from ._ffi.function import register_func
-from ._ffi.base import _RUNTIME_ONLY
-from . import api
-from . import tensor
-from . import schedule
-from . import expr
-from . import ir_pass
-from . import stmt as _stmt
-from . import container
-from . import module
-from . import codegen
-from . import ndarray
-from . import target as _target
-from . import make
-from ..devices import platform
-
-def replace_text(f_name, prev, new):
-    with open(f_name, 'r') as fp:
-        data = fp.read()
-    data = data.replace(prev, new)
-    with open(f_name, 'w') as fp:
-        fp.write(data)
-
-def run_process(cmd, pattern=None, env=None):
-    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
-    out, err = p.communicate()
-    if err: print("error raised: ", err.decode())
-    if pattern: return re.findall(pattern, out.decode("utf-8"))
-    return out.decode("utf-8")
-
-@register_func
-def tvm_callback_exec_evaluate(platform, mode, host_only):
-    # perform simulation and extract qor
-    qor = dict()
-
-    if platform == "vivado":
-      out = run_process("cd project; make vivado 2>&1")
-      print(out)
-
-    elif platform == "vivado_hls": 
-
-      assert os.system("which vivado_hls >> /dev/null") == 0, \
-        "cannot find vivado hls on system path"
-      ver = run_process("g++ --version", "\d\.\d\.\d")[0].split(".")
-      assert int(ver[0]) * 10 + int(ver[1]) >= 48, \
-        "g++ version too old {}.{}.{}".format(ver[0], ver[1], ver[2])
-
-      # for host only mode
-      if not os.path.isfile("project/kernel.cpp"):
-        replace_text("project/Makefile", "kernel.cpp", "")
-        replace_text("project/host.cpp", "#include \"kernel.h\"", "")
-
-      cmd = "cd project; make "
-      if mode == "sw_sim": cmd += "csim"
-      else: assert False
-
-      out = run_process(cmd + " 2>&1")
-      runtime = [k for k in out.split("\n") if "seconds" in k][0]
```

Comment:
I would suggest to move all these execution related part into a separate `runtime.py`